### PR TITLE
Customize ISO

### LINKED ISSUE #38: Customize ISO (sensor sensitivity)
From #21 focusing only on ISO configuration

### PR DESCRIPTION
**Pull request by friendoye**
Fixes #38 and also based on work of @hagabaka in #35.

### CODE REVIEW

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * Provides {@link Camera.Parameters} with methods to extract additional camera parameters.
+ */
+@SuppressWarnings("deprecation")
+public class RawCameraParametersProvider {
+
+    /* Raw camera params keys */
+    private static final String[] RAW_ISO_SUPPORTED_VALUES_KEYS = {
+            "iso-values", "iso-mode-values", "iso-speed-values", "nv-picture-iso-values"
+    };
+    private static final String[] RAW_ISO_CURRENT_VALUE_KEYS = {
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: this is actually unused. Are you sure that `applySensorSensitivity` will work across different devices?

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            if (rawValues != null) {
+                break;
+            }
+        }
+
+        return rawValues;
+    }
+
+    @Nullable
+    private String[] extractRawCameraValues(@NonNull String key) {
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: this has the same name with line 60 making it a bit hard to follow, let's make it more concrete like extract key/value

Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: Instead of `String[]` it's more solid to have a class KeyValue to return the pair or null is they don't exist

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Nullable
+    private String[] extractRawCameraValues(@NonNull String[] keys) {
+        String[] rawValues = null;
+
+        for (String key: keys) {
+            rawValues  = extractRawCameraValues(key);
+            if (rawValues != null) {
+                break;
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: you can early return here and return always null in line 70

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    public Set<Integer> getSensorSensitivityValues() {
+        final Set<Integer> isoValuesSet = new HashSet<>();
+
+        String[] rawValues = extractRawCameraValues(RAW_ISO_SUPPORTED_VALUES_KEYS);
+
+        // Should log that result is nullable or not?
+        if (rawValues != null) {
+            for (String value : rawValues) {
+                try {
+                    isoValuesSet.add(Integer.valueOf(value.trim()));
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: this exception can be avoided by using a custom class, check comment at the end of the file :)

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import android.support.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Provides {@link Camera.Parameters} with methods to extract additional camera parameters.
+ */
+@SuppressWarnings("deprecation")
+public class RawCameraParametersProvider {
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: we use RawCameraParametersProvider with either `getCameraParameters` or `getSensorSensitivityValues` in `CapabilitiesFactory`. I suggest we name this `CameraParametersProvider`, remove the `public getCameraParameters ` and have the previous parameter getters (e.g. `getSupportedPreviewSizes`) as public method of this class which is used in the `CapabilitiesFactory`

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/EmptyRange.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
+package io.fotoapparat.parameter.range;
+
+/**
+ * Null object for {@link Range}.
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/parameter/range/EmptyRange.java]: This is actually a range with no values (it's not a null object :) )

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java [hunk 7]
```diff
@@ -1,11 +1,39 @@
...
  */
 public class Ranges {
 
-    public static <T extends Comparable<T>> Range<T> range(T lowerBound, T upperBound) {
+    public static <T extends Comparable<T>> Range<T> range(@NonNull T lowerBound,
+                                                           @NonNull T upperBound) {
         return new ContinuousRange<>(lowerBound, upperBound);
     }
+
+    public static <T extends Comparable<T>> Range<T> discreteRange(@NonNull T value) {
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java]: this is creating a range with same high and low values, should be named so

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java [hunk 8]
```diff
@@ -1,11 +1,39 @@
...
+        return discreteRange(Collections.singletonList(value));
+    }
+
+    public static <T extends Comparable<T>> Range<T> discreteRange(@NonNull Collection<T> collection) {
+        List<T> list = (collection instanceof List)
+                ? (List<T>) collection
+                : new ArrayList<>(collection);
+        Collections.sort(list);
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java]: why the sorting? For the `toString`? Can be done in the constructor of DiscreteRange

File: fotoapparat/src/main/java/io/fotoapparat/parameter/selector/SensorSensitivitySelectors.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * @return {@link SelectorFunction} which selects default ISO value.
+     */
+    public static SelectorFunction<Range<Integer>,Integer> automaticSensorSensitivity() {
+        return new SelectorFunction<Range<Integer>, Integer>() {
+            @Override
+            public Integer select(Range<Integer> integerRange) {
+                return null;
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/parameter/selector/SensorSensitivitySelectors.java]: no default value?

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java (hunk 1 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: Hmm, will think about better organizing `applySensorSensitivity` work using `RAW_ISO_CURRENT_VALUE_KEYS`.

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java (hunk 5 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: Agreed. I'll work on it.

File: fotoapparat/src/main/java/io/fotoapparat/hardware/Capabilities.java [hunk 10]
```diff
@@ -92,6 +98,13 @@ public static Capabilities empty() {
     }
 
     /**
+     * @return the sensor sensitivity capability.
```
[the remark this answers is not in the document]
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/Capabilities.java]: Good point. Missed it :)

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java (hunk 2 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: Something like `extractKeyValue` or `extractRawCameraValue`? Didn't get it.

Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: Isn't it overkill for just getting raw string values from `cameraParameters`? Especially if `KeyValue` class, in my opinion, should have method `<T> T getValue(Class<T> clazz)`, that will be implemented with different checks on what returned value should be and then casting raw string to fit this class.

Comment (Diolor) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: Ah sorry, I think I didn't explain it very good and I just took a second look. This has been build with the purpose of reusuability, I now see 👍 . 

What I mainly think it's not nice is the `NumberFormatException` on a upper/later level.

So we can just make a nice logic only focusing to extract the ending `Set<Integer>` of ISOs and when we bump on the issue of duplicate/reusable code later on (because some similar behaviour with iso and splitting) we can think & solve it at that point.

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java (hunk 3 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/hardware/v1/RawCameraParametersProvider.java]: For me single return statement in method always better for understanding/readability, but OK.

File: fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java [hunk 11]
```diff
@@ -40,20 +43,21 @@
...
+    SelectorFunction<Collection<Size>, Size> previewSizeSelector = biggestSize();
+    SelectorFunction<Collection<FocusMode>, FocusMode> focusModeSelector = firstAvailable(
             continuousFocus(),
             autoFocus(),
             fixed()
     );
-    SelectorFunction<Flash> flashSelector = FlashSelectors.off();
-    SelectorFunction<Range<Integer>> previewFpsRangeSelector = Selectors.nothing();
+    SelectorFunction<Collection<Flash>, Flash> flashSelector = FlashSelectors.off();
+    SelectorFunction<Collection<Range<Integer>>, Range<Integer>> previewFpsRangeSelector = Selectors.nothing();
```
Comment (dmitry-zaitsev) [inline fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java]: If it is not a lot of effort for you, with new version of `SelectorFunction` it should be `SelectorFunction<Range<Integer>, Integer>` (same way as it is with `sensorSensitivitySelector`). Meaning is - we select one FPS from given range (not selecting one range from given collection of ranges).

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/DiscreteRange.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        this.values = valuesList;
+    }
+
+    @Override
+    public boolean contains(T value) {
+        return values.contains(value);
+    }
+
+    @Override
+    public T highest() {
```
Comment (dmitry-zaitsev) [inline fotoapparat/src/main/java/io/fotoapparat/parameter/range/DiscreteRange.java]: Makes sense to add `@Nullable` annotation.

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java [hunk 13]
```diff
@@ -1,11 +1,32 @@
...
+import java.util.Collection;
+
 /**
- * Class with factory methods for {@link Range}.
+ * Class with factory and utility methods for {@link Range}.
  */
 public class Ranges {
 
-    public static <T extends Comparable<T>> Range<T> range(T lowerBound, T upperBound) {
+    public static <T extends Comparable<T>> Range<T> range(@NonNull T lowerBound,
```
Comment (dmitry-zaitsev) [inline fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java]: JavaDocs are missing. Without them it is not entirely clear for the user of the API what is the difference between `range` and `discreteRange`.

File: fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java (hunk 11 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java]: Almost. Since we're using `setPreviewFpsRange(min, max)`, we should provide 2 values. Furthermore, we're losing flexibility in Fotoapparat API.

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/DiscreteRange.java (hunk 12 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/parameter/range/DiscreteRange.java]: Sure.

File: fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java (hunk 11 above)
Comment (dmitry-zaitsev) [reply fotoapparat/src/main/java/io/fotoapparat/FotoapparatBuilder.java]: Ah, right. Makes sense. Let's keep it as it is.

File: fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java (hunk 13 above)
Comment (friendoye) [reply fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java]: Maybe replace `range` with `continuousRange` and refer to `Range`'s classes JavaDocs?

Comment (dmitry-zaitsev) [reply fotoapparat/src/main/java/io/fotoapparat/parameter/range/Ranges.java]: Would be even better.

**Comment by friendoye** [comment]
@Diolor could you review my changes again? Thanks in advance.

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/CameraParametersDecorator.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+import android.hardware.Camera.Size;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Decorator for {@link Camera.Parameters} with methods
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/CameraParametersDecorator.java]: 🎉

File: fotoapparat/src/main/java/io/fotoapparat/hardware/v1/CameraParametersDecorator.java [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+    /**
+     * Returns set of ISO values, that camera supports.
+     *
+     * @return the set of supported ISO values.
+     */
+    @NonNull
+    public Set<Integer> getSensorSensitivityValues() {
+        String[] rawValues = extractRawCameraValues(RAW_ISO_SUPPORTED_VALUES_KEYS);
+        return convertParamsToInts(rawValues);
```
Comment (Diolor) [inline fotoapparat/src/main/java/io/fotoapparat/hardware/v1/CameraParametersDecorator.java]: I think that's good too instead of custom classed 👍

**Comment by Diolor** [comment]
Thanks guys! @friendoye @hagabaka
